File: go/pdfcontext/infra_ghostscript.go
package pdfcontext

import (
	"os"
	"os/exec"
	"path/filepath"
	"strconv"
	"time"

	"pdftoolkit/sharedkernel"
)

// GhostscriptCompressor is the escalation/baseline engine: binary-searches
// -dColorImageResolution to sit just under the byte budget. Mirrors the Python
// adapter. External `gs` dependency confined here.
type GhostscriptCompressor struct {
	Bin        string
	Iterations int
}

func NewGhostscriptCompressor() GhostscriptCompressor {
	return GhostscriptCompressor{Bin: "gs", Iterations: 8}
}

func (g GhostscriptCompressor) Name() string { return "ghostscript" }

func (g GhostscriptCompressor) run(src, dst string, dpi int) (int64, error) {
	d := strconv.Itoa(dpi)
	args := []string{
		"-sDEVICE=pdfwrite", "-dNOPAUSE", "-dQUIET", "-dBATCH", "-dPDFSETTINGS=/default",
		"-dColorImageResolution=" + d, "-dGrayImageResolution=" + d, "-dMonoImageResolution=" + d,
		"-dDownsampleColorImages=true", "-dDownsampleGrayImages=true",
		"-sOutputFile=" + dst, src,
	}
	if err := exec.Command(g.Bin, args...).Run(); err != nil {
		return 0, err
	}
	fi, err := os.Stat(dst)
	if err != nil {
		return 0, err
	}
	return fi.Size(), nil
}

func (g GhostscriptCompressor) Compress(src sharedkernel.MediaFile, t CompressionTarget, out string) (CompressionResult, error) {
	start := time.Now()
	ceiling := t.Budget.Ceiling()
	srcSize := src.SizeBytes()

	res := func(dpi int) CompressionResult {
		var after int64
		if fi, err := os.Stat(out); err == nil {
			after = fi.Size()
		}
		return CompressionResult{
			Output: sharedkernel.MediaFile{Path: out}, Engine: g.Name(),
			BeforeBytes: srcSize, AfterBytes: after, DpiUsed: dpi,
			ElapsedMs: time.Since(start).Milliseconds(),
		}
	}

	if srcSize <= ceiling { // already fits
		if err := copyFile(src.Path, out); err != nil {
			return CompressionResult{}, err
		}
		return res(0), nil
	}

	lo, hi := t.DpiRange[0], t.DpiRange[1]
	best := lo
	tmp := filepath.Join(os.TempDir(), "pdftoolkit_gsprobe.pdf")
	for i := 0; i < g.Iterations; i++ {
		mid := (lo + hi) / 2
		size, err := g.run(src.Path, tmp, mid)
		if err != nil {
			return CompressionResult{}, err
		}
		if size <= ceiling {
			best = mid
			lo = mid
		} else {
			hi = mid
		}
	}
	_ = os.Remove(tmp)
	if _, err := g.run(src.Path, out, best); err != nil {
		return CompressionResult{}, err
	}
	// Hard rule #1: never emit larger than input.
	if fi, err := os.Stat(out); err == nil && fi.Size() >= srcSize {
		if err := copyFile(src.Path, out); err != nil {
			return CompressionResult{}, err
		}
		best = 0
	}
	return res(best), nil
}

func copyFile(src, dst string) error {
	b, err := os.ReadFile(src)
	if err != nil {
		return err
	}
	return os.WriteFile(dst, b, 0o644)
}

File: rust/src/pdf_context/infra_ghostscript.rs
//! pdf context — Ghostscript engine (escalation/baseline). Uses std::process only,
//! so no external crates. Mirrors the Python and Go GS adapters.

use std::fs;
use std::path::Path;
use std::process::Command;
use std::time::Instant;

use super::domain::{CompressionResult, CompressionTarget};
use super::ports::Compressor;
use crate::shared_kernel::MediaFile;

pub struct GhostscriptCompressor {
    pub bin: String,
    pub iterations: u32,
}

impl Default for GhostscriptCompressor {
    fn default() -> Self {
        Self { bin: "gs".into(), iterations: 8 }
    }
}

impl GhostscriptCompressor {
    fn run(&self, src: &Path, dst: &Path, dpi: u32) -> Result<u64, String> {
        let status = Command::new(&self.bin)
            .args([
                "-sDEVICE=pdfwrite",
                "-dNOPAUSE",
                "-dQUIET",
                "-dBATCH",
                "-dPDFSETTINGS=/default",
                &format!("-dColorImageResolution={dpi}"),
                &format!("-dGrayImageResolution={dpi}"),
                &format!("-dMonoImageResolution={dpi}"),
                "-dDownsampleColorImages=true",
                "-dDownsampleGrayImages=true",
                &format!("-sOutputFile={}", dst.display()),
            ])
            .arg(src)
            .status()
            .map_err(|e| e.to_string())?;
        if !status.success() {
            return Err("ghostscript failed".into());
        }
        Ok(fs::metadata(dst).map_err(|e| e.to_string())?.len())
    }
}

impl Compressor for GhostscriptCompressor {
    fn name(&self) -> &str {
        "ghostscript"
    }

    fn compress(
        &self,
        source: &MediaFile,
        target: &CompressionTarget,
        out: &Path,
    ) -> Result<CompressionResult, String> {
        let start = Instant::now();
        let ceiling = target.budget.ceiling();
        let src_size = source.size_bytes();

        let make = |dpi: u32| -> CompressionResult {
            let after = fs::metadata(out).map(|m| m.len()).unwrap_or(0);
            CompressionResult {
                output: MediaFile::new(out),
                engine: "ghostscript".into(),
                before_bytes: src_size,
                after_bytes: after,
                dpi_used: dpi,
                quality_used: 0,
                score: None,
                elapsed_ms: start.elapsed().as_millis(),
                escalated: false,
            }
        };

        if src_size <= ceiling {
            fs::copy(&source.path, out).map_err(|e| e.to_string())?;
            return Ok(make(0));
        }

        let (mut lo, mut hi) = target.dpi_range;
        let mut best = lo;
        let tmp = std::env::temp_dir().join("pdftoolkit_gsprobe.pdf");
        for _ in 0..self.iterations {
            let mid = (lo + hi) / 2;
            let size = self.run(&source.path, &tmp, mid)?;
            if size <= ceiling {
                best = mid;
                lo = mid;
            } else {
                hi = mid;
            }
        }
        let _ = fs::remove_file(&tmp);
        self.run(&source.path, out, best)?;
        // Hard rule #1: never emit larger than input.
        if fs::metadata(out).map(|m| m.len()).unwrap_or(0) >= src_size {
            fs::copy(&source.path, out).map_err(|e| e.to_string())?;
            best = 0;
        }
        Ok(make(best))
    }
}
